### DISCUSSION TITLE
Frozen string literal true

### DISCUSSION THREAD

**Discussion by B3Y0N4** [19/classroom#5] [Ruby] [answered]
1. Soru
#frozen_string_literal: true kullandığımda
str = ' '
str <<  'merhaba'
p str
Bu şekilde yaptığımda   can't modify frozen String: "" (FrozenError)  hatasını alıyorum ki olması gereken fakat
#frozen_string_literal: true
str = String.new ' ' 
str << 'merhaba'
p str Bu şekilde hata vermiyor yada 
#frozen_string_literal: true
str = ''
str += 'merhaba'
p str 
Bu şekilde hata vermiyor ve sonuçta dizgiyi değiştirdik 
Neden bu şekilde çalışıyor?

2. Soru
(Dikkat! String.new’i argümansız çalıştırırsanız karakter kodlaması
ASCII oluyor)
Bu cümleyi açabılırmisıniz tam olarak ne demek istiyor

**Comment by roktas** [comment, marked as answer]
**1.a.** `String.new ''` yaptığında üretilen string "literal" değil "dinamik" bir string, yani `frozen_string_literal: true`'nun kapsamı dışında.

**1.b.** `str += 'merhaba'` ile `str` literal string'i değiştirmiyorsun, yeni (ve artık frozen olmayan) bir string yaratıyorsun.  hint: `str = str + 'merhaba'`

**2.** Şöyle anlatayım:

```ruby
s = String.new
s.encoding #=> #<Encoding:ASCII-8BIT>
s = String.new ''
s.encoding #=> #<Encoding:UTF-8>
```

**Reply by B3Y0N4** [reply]
1.b. 
frozen_string_literal: true yaptiğmizda dizgilerin değerleri değiştirilemez (immutable) olmiyor mu ? 
#frozen_string_literal: true
str = 'Merhaba ' #ör: str değeri burda merhaba 
str += 'Dunya'
p str # Ama burda Merhaba Dunya 
yanı bir string üzerinde yeni bir string yaratıyorsak özgun string değiştrımiş olmıyormuyuz?.

**Reply by roktas** [reply]
Hayır, tanımlayıcılara (identifier) fazla anlam yüklüyorsun.  "Değiştirilemezlik" konusu veriler için geçerli.  Immutable olan filan satırda `str` etiketiyle (identifier) eriştiğin bellek bölgesindeki `'Merhaba'` verisi.  Yani `new_str = str + 'Merhaba'` ile `str = str + 'Merhaba'` arasında bu açıdan bir fark yok, sadece identifier'ı "recycle" etmişsin.

**Reply by roktas** [reply]
Şöyle de bak:

```ruby
s = "bir string"
s.object_id #=> kabaca bellekteki bölgeyi temsil eden bir sayı (her makinede ve/veya oturumda farklı çıkar)
s += " ekleme"
s.object_id #=> kabaca bellekteki bölgeyi temsil eden bir sayı (her makinede ve/veya oturumda farklı çıkar)
```

Örnekte iki satırdaki `object_id`'lerin farklı olacağını göreceksin.  Yani ikinci eklemede bellekte yeni bir string oluşturuluyor (immutable olan eski veriye dokunulmuyor), "ama işe bakın ki" bu veriye de aynı isimle (`s`) erişiyor oluyoruz.

**Reply by roktas** [reply]
https://github.com/roktas/classroom/discussions/10#discussioncomment-2028218 cevabında ufak düzenlemeler yaptım, tekrar okumayı unutmayın.

**Comment by roktas** [comment]
`frozen_string_literal: true` direktifiyle etkinleştirdiğiniz "değiştirilemez (dondurulmuş) dizgi (string) sabitleri" neye yarıyor?  Hemen akla gelenler:

- Kaynak koda gömdüğünüz sabit stringlerde eşdeğer her string içeriğine aynı bellek alanından erişiyorsunuz.  Veri tekrarı yok, bellek daha etkin kullanılmış oluyor (ki bu da GC'nin yapacağı işi azaltır).
- Değiştirilmemesi gereken verilerin kaza eseri de olsa değiştirilmesini engellemiş oluyorsunuz.  Bu sayede ilk bakışta hemen görülemeyecek bazı hatalar da kod çalıştırılmadan farkedilmiş oluyor.  İnceleyin:

```ruby
def hatali_buyut_beni_metodu(string)
  string.upcase!
end

s = 'bu sabit'
s.freeze # frozen_string_literal: true ile aynı işi yapar, IRB'de denemeniz için ekliyorum

hatali_buyut_beni_metodu(s) # ERROR

def buyut_beni_metodu(string)
  string.upcase
end

buyut_beni_metodu(s) # OK
```

**Reply by B3Y0N4** [reply]
tamamdir hocam teşşekür ederim
